### DISCUSSION TITLE
Setting headers for dynamically-created pages

### DISCUSSION THREAD

**Discussion by decaffeinatedio** [rubycdp/ferrum#411] [Q&A]
I'd like to be able to set custom headers and read network traffic when a new page is dynamically created, e.g. via a mouse click, rather than explicitly created by my application with `create_page`. 

I'm trying to listen for `.on('Target.targetCreated')` events, but I believe these occur too late in the process, after headers can be meaningfully overwritten for the initial request.

**Comment by decaffeinatedio** [comment]
A little more information: 

I was able to stop tabs from automatically loading by calling `browser.command('Target.setAutoAttach', waitForDebuggerOnStart: true, autoAttach: true, flatten: true)` and then patching in the following in the `prepare_page` method in `Page` to complete loading:

```
    def prepare_page
      command("Page.enable")
      command("Runtime.enable")
--> command("Runtime.runIfWaitingForDebugger") <---
      command("DOM.enable")
      command("CSS.enable")
      command("Log.enable")
      command("Network.enable")

      if use_authorized_proxy?
        network.authorize(user: @pr...
```

However, this _only_ works for `window.open`. For normal clicks, the tab is stuck in a loading state and never (seemingly) receives the `Runtime.runIfWaitingForDebugger` signal. 

Relevant Puppeteer discussion: https://github.com/puppeteer/puppeteer/issues/3667

**Comment by route** [comment]
I think it's now addressed in master. But your original title says something about headers, but the description is about flatten mode. Hope it resolves all the issues :)
